aws/lambda/rds-proxy/lambda_function.py

```diff
@@ -0,0 +1,298 @@
+"""
+import json
+import datetime
+import os
+from typing import *
+from contextlib import closing
+import re
+
+
+import mysql.connector
+
+
+_connections = {
+    "reader": {
+        "connection": None,
+        "user": "db_user",
+        "password": "db_password",
+        "database": "pytorch",
+    },
+    "inserter": {
+        "connection": None,
+        "user": "db_user_inserter",
+        "password": "db_password_inserter",
+        "database": "metrics",
+    },
+    "creator": {
+        "connection": None,
+        "user": "db_user_creator",
+        "password": "db_password_creator",
+        "database": "metrics",
+    },
+}
+
+
+def get_connection(name: str):
+    field = _connections[name]
+
+    if field["connection"] is None:
+        field["connection"] = mysql.connector.connect(
+            host=os.environ["db_host"],
+            port=3306,
+            user=os.environ[field["user"]],
+            password=os.environ[field["password"]],
+            database=field["database"],
+        )
+
+    return field["connection"]
+
+
+SAVED_QUERIES = {"sample": "select name from workflow_run limit 10"}
+
+TYPE_MAP = {
+    "int": "INTEGER",
+    "string": "VARCHAR(300)",
+}
+
+
+NAME_REGEX = re.compile("^[a-z_]+$")
+
+
+def validate_schema_name(s: str):
+    if NAME_REGEX.match(s) is not None:
+        return s
+    else:
+        raise RuntimeError(f"Invalid name: {s}")
+
+
+def safe_join(s: Union[str, List[str]], join_str: str = ", ") -> str:
+    if isinstance(s, str):
+        s = [s]
+
+    return join_str.join([validate_schema_name(x) for x in s])
+
+
+def build_query(body):
+    # If the request is a simple query we can just build it manually rather
+    # than having to hard code it in the list above
+    params = []
+    table_name = validate_schema_name(body["table_name"])
+
+    query = f"SELECT {safe_join(body['fields'])} FROM {safe_join(table_name)}"
+
+    where = body.get("where", None)
+    if where is not None:
+        if not isinstance(where, list):
+            where = [where]
+        for item in where:
+            item["field"] = validate_schema_name(item["field"])
+        query += " WHERE"
+        items = [f"{n['field']} {'like' if n['like'] else '='} %s" for n in where]
+        query += f" {' and '.join(items)}"
+        params += [n["value"] for n in where]
+
+    group_by = body.get("group_by", None)
+    if group_by is not None:
+        query += f" GROUP BY {safe_join(group_by)}"
+
+    order_by = body.get("order_by", None)
+    if order_by is not None:
+        query += f" ORDER BY {safe_join(order_by)}"
+
+    limit = body.get("limit", None)
+    if limit is not None:
+        query += f" LIMIT %s"
+        params.append(int(limit))
+
+    return query, params
+
+
+def run_query(query: str, params: List[str], connection: Any) -> List[Dict[str, str]]:
+    print(f"Executing '{query}' with params: {params}")
+    if "--" in query:
```

Comment:
are there other validations we should do or will the mysql connector handle it?

Reply:
The ones here are pretty basic just to avoid anything egregious, the mysql doesn't let you parameterize things like table names or fields so we're stuck setting them manually afaik, any user provided data should only be ascii + numbers + `_` so we should be mostly okay